src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+}
+
+// Try including a node's input at specified index into current statement
+bool VectorMaskedLoop::collect_statements_helper(
+          Node* node, uint idx, Node_List* stmt, Node_List* worklist) {
+  Node* in = node->in(idx);
+  if (stmt->contains(in) || !in_body(in)) {
+    // Input is already included in current statement or out of loop
+    return true;
+  }
+
+  // Check support for special inputs first and then general ones
+  if (is_loop_iv_or_incr(in)) {
+    // 1) Check the support of loop iv or increment node input
+    BasicType bt = statement_bottom_type(stmt);
+    bt = is_subword_type(bt) ? bt : T_INT;
+    if (VectorNode::is_populate_index_supported(bt)) {
+      return true;
+    } else {
+      trace_msg(in, "Populate index operation is not supported");
+      return false;
+    }
+  } else if (in->is_Phi()) {
+    // 2) We don't support phi nodes except the iv phi of the loop
+    trace_msg(in, "Found unsupported phi input");
+    return false;
+  } else if (in->is_Load()) {
+    // 3) Ok to include a load node if it's supported memory access
+    if (supported_mem_access(in->as_Load())) {
+      stmt->push(in);
+      return true;
+    } else {
+      trace_msg(in, "Found unsupported memory load input");
+      return false;
+    }
+  } else if (VectorNode::is_shift(in) && in_body(in->in(2))) {
+    // 4) We don't support shift operations with variant shift count
+    trace_msg(in, "Variant shift count is not supported");
+    return false;
+  } else {
+    // 5) For other general inputs, include it and also push it into the
+    //    worklist to collect inputs recursively
+    worklist->push(in);
+    stmt->push(in);
+    return true;
+  }
+}
+
+// Collect lists of nodes that make up loop statements
+bool VectorMaskedLoop::collect_statements() {
+  // First, initialize each statement from a store node.
+  for (int idx = 0; idx < _body_nodes.length(); idx++) {
+    Node* node = _body_nodes.at(idx);
+    if (node->is_Store() && supported_mem_access(node->as_Store())) {
+      // Create a new statement and add the store into its node list
+      Node_List* stmt = new Node_List(_arena);
+      stmt->push(node);
+      _stmts.append(stmt);
+    }
+  }
+
+  // Do early bail out if no statement is created
+  int num_stmts = _stmts.length();
+  if (num_stmts == 0) {
+    trace_msg(nullptr, "No vectorizable statement is found");
+    return false;
+  }
+
+  // Then, extend each statement by recursively including input nodes. Bail out
+  // if unsupported node is found.
+  for (int idx = 0; idx < num_stmts; idx++) {
+    Node_List* stmt = _stmts.at(idx);
+    assert(stmt->size() == 1, "Each statement should have exactly one node");
+    Node* store = stmt->at(0);
+    // Add value input of the store node into a worklist to include more nodes
+    // into current statement
+    Node_List* worklist = new Node_List(_arena);
+    if (!collect_statements_helper(store, MemNode::ValueIn, stmt, worklist)) {
+      return false;
```

Comment:
Why does the `store` need special handling here? Can you not just throw it on the `worklist`? Would be nice to have the code be shorter ;)

Reply:
Good catch! Fixed.